flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffleProducer.java

```diff
@@ -0,0 +1,197 @@
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.serialization.TypeInformationSerializationSchema;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.api.java.functions.KeySelector;
+import org.apache.flink.core.memory.DataOutputSerializer;
+import org.apache.flink.runtime.state.KeyGroupRangeAssignment;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.connectors.kafka.FlinkKafkaException;
+import org.apache.flink.streaming.connectors.kafka.FlinkKafkaProducer;
+import org.apache.flink.util.Preconditions;
+import org.apache.flink.util.PropertiesUtil;
+
+import org.apache.kafka.clients.producer.ProducerRecord;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Properties;
+
+import static org.apache.flink.streaming.connectors.kafka.shuffle.FlinkKafkaShuffle.PARTITION_NUMBER;
+
+/**
+ * Flink Kafka Shuffle Producer Function.
+ * It is different from {@link FlinkKafkaProducer} in the way handling elements and watermarks
+ */
+@Internal
+public class FlinkKafkaShuffleProducer<IN, KEY> extends FlinkKafkaProducer<IN> {
+	private final KafkaSerializer<IN> kafkaSerializer;
+	private final KeySelector<IN, KEY> keySelector;
+	private final int numberOfPartitions;
+
+	FlinkKafkaShuffleProducer(
+		String defaultTopicId,
+		TypeInformationSerializationSchema<IN> schema,
+		Properties props,
+		KeySelector<IN, KEY> keySelector,
+		Semantic semantic,
+		int kafkaProducersPoolSize) {
+		super(defaultTopicId, (element, timestamp) -> null, props, semantic, kafkaProducersPoolSize);
+
+		this.kafkaSerializer = new KafkaSerializer<>(schema.getSerializer());
+		this.keySelector = keySelector;
+
+		Preconditions.checkArgument(
+			props.getProperty(PARTITION_NUMBER) != null,
+			"Missing partition number for Kafka Shuffle");
+		numberOfPartitions = PropertiesUtil.getInt(props, PARTITION_NUMBER, Integer.MIN_VALUE);
+	}
+
+	/**
+	 * This is the function invoked to handle each element.
+	 * @param transaction transaction state;
+	 *                    elements are written to Kafka in transactions to guarantee different level of data consistency
+	 * @param next element to handle
+	 * @param context context needed to handle the element
+	 * @throws FlinkKafkaException for kafka error
+	 */
+	@Override
+	public void invoke(KafkaTransactionState transaction, IN next, Context context) throws FlinkKafkaException {
+		checkErroneous();
+
+		// write timestamp to Kafka if timestamp is available
+		Long timestamp = context.timestamp();
+
+		int[] partitions = getPartitions(transaction);
+		int partitionIndex;
+		try {
+			partitionIndex = KeyGroupRangeAssignment
+				.assignKeyToParallelOperator(keySelector.getKey(next), partitions.length, partitions.length);
+		} catch (Exception e) {
+			throw new RuntimeException("Fail to assign a partition number to record");
+		}
+
+		ProducerRecord<byte[], byte[]> record = new ProducerRecord<>(
```

Comment:
I think during my first test, the timestamp read from ConsumerRecord is always the timestamp when it was created. 
But i did not spend time looking into this because as stated above, the SerDe part can be generalized. it is not just Kafka. So i would like to make the SerDe less dependant on Kafka.

Reply:
Okay seems plausible.